### ISSUE TITLE
Support for Markdown editing and previews

### ISSUE THREAD

**Issue by dariuszparys** [OpenLiveWriter/OpenLiveWriter#11]
OLW needs support for Markdown editing and preview. Will start looking at implementing it based on Marked (something that VSCode is using).

@dend

**Comment by MaximRouiller** [comment]
Nothing about CommonMark?

**Comment by ahmetalpbalkan** [comment]
+1

**Comment by tugberkugurlu** [comment]
:+1:

**Comment by thomaslevesque** [comment]
:+1:

**Comment by dend** [comment]
I started to work on a variation of this based on [Marked](https://github.com/chjj/marked).

**Comment by DDReaper** [comment]
Wow, definitely 😍

**Comment by caioproiete** [comment]
:+1:

**Comment by JohnBlood** [comment]
I agree. Markdown is important for the modern web. Makes writing _easier_ and **faster**.

**Comment by andrei-markeev** [comment]
:+1: code highlight **wanted dearly** :)

**Comment by KirinHuang** [comment]
**Wanted badly!**

**Comment by dhusemann** [comment]
textmate outlining files support??

**Comment by usagi** [comment]
:+1:

**Comment by onovotny** [comment]
Question here for discussion: for Markdown, should OLW publish the rendered HTML or the raw markdown to blog engines that support it?

The recent versions of WordPress support Markdown with the WP-Markdown plugin. Of course that leaves the fidelity issue where the blog's rendering may differ from the OLW preview.

Thoughts?

**Comment by writeline** [comment]
If blogs support both, then I think this should be settable by the user with a default set by the blog provider in use. For instance I would assume Wordpress to accept HTML natively where Ghost would take Markdown.

**Comment by smiley** [comment]
It should support both modes to accommodate blog engines like Jekyll. For example, OLW could post the raw Markdown file to the `_posts` folder in your repository, leaving the rest to your Jekyll workflow. (E.g.: GitHub Pages)

**Comment by dend** [comment]
What @writeline mentioned. My hunch is that we need to do some sort of early detection to check whether we need to push either MD or HTML.

**Comment by DDReaper** [comment]
I'd agree, having both options make sense, especially since one is just an extension of the other.

**Comment by onovotny** [comment]
My bigger concern around having both as an option would be around if OLW supported things like code-highlighting MD extensions that some providers don't. It'd look great locally but then would render differently.

**Comment by dend** [comment]
Not to randomize the task set, I am trying to make sure that we get proper markdown rendering and push done as Pri-0. Highlighting and the rest are "nice to have" extras that we can tackle later.

**Comment by onovotny** [comment]
@dend agree re priorities. On that "nice to have" list might also be a way to specify that certain MD features should be translated to HTML while leaving the others alone.

For example:
```
```csharp 
```
could be translated to html inline while leaving the rest as markdown for readability

**Comment by shanselman** [comment]
What is the experience going to look like, UXwise @writeline? You'll go to the source tab and see Markdown, then see the preview? You'll completely hide the HTML?

**Comment by dend** [comment]
The way I am looking at it, there is:

* **One-way Markdown** - User composes MD in the editor with no additional help, in a format similar to what VSCode has today. Then going to Preview will generate you a preview of what you just typed.

* **Two-way Markdown** - User writes regular text, with all the formatting goodness. The MD source is generated automatically in this case and so is the preview based on the auto-generated source.

Right now I am working on the first version (one-way Markdown). User swaps between Markdown and HTML modes in the ribbon with no auto-detection of what the blog supports yet. Once the basic infra is working, I think we can push for a more advanced mechanism.

Thoughts?

**Comment by rebornix** [comment]
@dend that would be great. Supporting writing MD in editor will unblock a lot of issues as current editor will transform text to HTML directly. 

Once MD authoring is enabled, we can publish correct content to both GitHub Page, Ghost, etc. Then people can send out their PR about MD related blog provider support.

**Comment by DDReaper** [comment]
VSCode effectively works the same way as option 1.
Having the 2-way would be nice but could be tricky given all the options available.
I'd say get it out the door with one-way first, then add two way later if there's demand for it.

**Comment by MaximRouiller** [comment]
* Option 1: Markdown Preview
* Option 2: Markdown Editor 

That's the actual name that are being used in the wild right now.

**Comment by writeline** [comment]
@shanselman I would definitely leave source and preview as it is. So the `edit` tab could be used to write markdown. That would mean that we need a mode to switch between HTML and Markdown editing. The question is what happens with all those ribbon commands which users are using to edit posts. They could still function in markdown mode too, producing different results based on the `edit` tab mode. But I haven't looked into how commands are implemented currently, it could be a bigger task than initially thought.

This is how I think it could work:

- Introduce a switch for editing HTML or Markdown
- Leverage the `edit` tab as editor pane for both. It shows the caption `Edit - HTML Mode` when in HTML mode and `Edit - Markdown` when in Markdown mode.
- In Markdown mode a completely new ribbon bar appears with all commands supported by http://daringfireball.net/projects/markdown/syntax
- In Markdown mode source works only uni-directional to keep it simple
- Integrate an image upload mechanism for all local added images (if the blog provider supports image hosting)

Any thoughts on this?

**Comment by MaximRouiller** [comment]
If you are going with the Daringfireball syntax... good luck. The author basically dropped "the spec" and many different libraries will render different things at different scenarios. 

We might end up with lots of different bug depending on who test with what. 

The only "Markdown" syntax that has unit test, properly defined syntax is CommonMark.

**Comment by dend** [comment]
@writeline that's actually how it works today - ribbon switch between HTML and Markdown.

@MaximRouiller not worried about the spec at this point. Integrating Marked into the product directly.

**Comment by MaximRouiller** [comment]
Why `marked`?

**Comment by dend** [comment]
At this point simpler to integrate. VSCode uses the same engine.

**Comment by MaximRouiller** [comment]
Fair enough.

I would highly recommend checking out CommonMark however. The presence of a spec for something as widely changing as Markdown makes it easier to know if we are doing things correctly or "good enough".

Converting Markdown to HTML with CommonMark is quite easy.

C#:

    var html = CommonMarkConverter.Convert(markdown);

JavaScript:

    var reader = new commonmark.Parser();
    var writer = new commonmark.HtmlRenderer();
    var parsed = reader.parse("markdown");
    var html = writer.render(parsed);

**Comment by rebornix** [comment]
IMHO, to most Markdown users, if the dialect has GFM support, then it's a good one.

**Comment by rebornix** [comment]
@dend any update on this task?

**Comment by dend** [comment]
@rebornix Ironing out a couple of bugs. Other than that, looks good, should check in shortly.

**Comment by jmatthiesen** [comment]
@dend I'll jump in with the requests here. :) How's the Markdown support looking? Any early bits available in a branch where you could use some extra eyes?

**Comment by Eucaly** [comment]
Hi all,
I really love open live writer (since 2007 Windows Live Writer Beta).

Share my work [bZmd](https://github.com/beZong/bZmd), ultilizing c# markdown preview with different markdown engines, hope that helps.
* [Strike.IE](https://github.com/SimonCropp/Strike) -- to render GitHub-flavored markdown offline
* [GitHub markdown API](http://developer.github.com/v3/markdown) -- require network connection
* MarkdownSharp ...

I've packed them to follow generic [IMarkdown](https://github.com/beZong/bZmd/tree/master/bZmd/myMarkdown) interface class. 
```
	interface IMarkdown
	{
		string FileToHtml(FileInfo info);
		string StringToHtml(string sIn);
		void FileToHtml(FileInfo info, Stream CssFile, out string sOut);
		string FileToHtml(FileInfo info, Stream CssFile);
	}
```

With some further work, it should be possible to 
* fallback when no network connection
* Allow user to select different render backend

**Comment by jongalloway** [comment]
Adding another more recent contender: https://github.com/lunet-io/markdig

**Comment by huoxudong125** [comment]
+1

**Comment by shanselman** [comment]
Thanks for reporting this and for helping Open Live Writer over the years. We are doing an issue-bankruptcy cleanup so we can make the repository maintainable again and focus on current, actionable work. This older issue is being closed as part of that cleanup, not because it was unimportant. If this still reproduces in the current code or is still relevant to the revitalization work, please open a fresh issue with current details and we will triage it going forward.
